client/src/main/java/org/eclipse/hono/client/impl/MappingAndDelegatingCommandHandler.java

```diff
@@ -0,0 +1,288 @@
+        } else {
+            // command is invalid, try to find local command handler to let it reject the command (and also report metrics for that), otherwise reject it directly
+            commandTargetMapper.getTargetGatewayAndAdapterInstance(tenantId, deviceId, spanContext)
+                    .setHandler(ar -> {
+                        CommandHandlerWrapper commandHandler = null;
+                        if (ar.succeeded()) {
+                            final String targetDevice = ar.result().getString(DeviceConnectionConstants.FIELD_PAYLOAD_DEVICE_ID);
+                            final String targetAdapterInstance = ar.result().getString(DeviceConnectionConstants.FIELD_ADAPTER_INSTANCE_ID);
+                            if (adapterInstanceId.equals(targetAdapterInstance)) {
+                                final String gatewayId = !deviceId.equals(targetDevice) ? targetDevice : null;
+                                commandHandler = adapterInstanceCommandHandler
+                                        .getDeviceSpecificCommandHandler(tenantId, deviceId, gatewayId);
+                            }
+                        }
+                        if (commandHandler != null) {
+                            // let the device specific handler reject the command
+                            commandHandler.handleCommand(commandContext);
+                        } else {
+                            LOG.debug("command message is invalid: {}", command);
+                            commandContext.reject(new ErrorCondition(Constants.AMQP_BAD_REQUEST,
+                                    "malformed command message"));
+                        }
+                    });
+
+        }
+    }
+
+    private void mapAndDelegateIncomingCommand(final CommandContext originalCommandContext) {
+        final Command originalCommand = originalCommandContext.getCommand();
+        if (!originalCommand.isValid()) {
+            originalCommandContext.reject(new ErrorCondition(Constants.AMQP_BAD_REQUEST, "malformed command message"));
+            return;
+        }
+        final String tenantId = originalCommand.getTenant();
+        final String originalDeviceId = originalCommand.getDeviceId();
+        // determine last used gateway device id
+        LOG.trace("determine command target gateway/adapter for [{}]", originalCommand);
+        final Future<JsonObject> commandTargetFuture = commandTargetMapper.getTargetGatewayAndAdapterInstance(tenantId, originalDeviceId,
+                originalCommandContext.getTracingContext());
+
+        commandTargetFuture.setHandler(commandTargetResult -> {
+            if (commandTargetResult.succeeded()) {
+                final String targetDeviceId = commandTargetResult.result().getString(DeviceConnectionConstants.FIELD_PAYLOAD_DEVICE_ID);
+                final String targetAdapterInstance = commandTargetResult.result().getString(DeviceConnectionConstants.FIELD_ADAPTER_INSTANCE_ID);
+
```

Comment:
this is the same code as in the main method, isn't it? would it be possible to change the behavior so that we first try to delegate processing of a command to a local handler in any case (regardless of whether a command is valid or not) and then, if valid, forward it to the correct adapter instance?

Reply:
Good point. I've refactored the method here and made that change.